File: Interpretadas/Python/quick_sort.py
import sys
import time
import random
import locale

# Classe para armazenar as métricas
class Metricas:
    def __init__(self):
        self.comparacoes = 0
        self.trocas = 0
        self.tempoExecucao = 0.0  # Armazena o tempo de execução em segundos
        self.memoriaUsada = 0  # Armazena a quantidade de memória usada em bytes

def mediana_de_tres(arr, low, high):
    mid = (low + high) // 2
    if arr[low] > arr[mid]:
        arr[low], arr[mid] = arr[mid], arr[low]
    if arr[low] > arr[high]:
        arr[low], arr[high] = arr[high], arr[low]
    if arr[mid] > arr[high]:
        arr[mid], arr[high] = arr[high], arr[mid]
    # Coloca o pivot (mediana) no final
    arr[mid], arr[high] = arr[high], arr[mid]
    return arr[high]

# Implementação do Quick Sort com mediana de três
def partition(arr, low, high, metricas):
    pivot = mediana_de_tres(arr, low, high)
    metricas.memoriaUsada += 4  # Considerando 4 bytes por int para o pivot
    i = low - 1

    for j in range(low, high):
        metricas.comparacoes += 1
        if arr[j] <= pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]
            metricas.trocas += 1

    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    metricas.trocas += 1
    return i + 1

def quickSort(arr, low, high, metricas):
    if low < high:
        pi = partition(arr, low, high, metricas)
        quickSort(arr, low, pi - 1, metricas)
        quickSort(arr, pi + 1, high, metricas)


def carregarVetor(tamanho, caso):
    with open('vetores/vetores_input.txt', 'r') as f:
        lines = f.readlines()
        key = f"{tamanho} {caso}:"
        for i in range(len(lines)):
            if lines[i].strip() == key:
                vetor = list(map(int, lines[i + 1].strip().split()))
                return vetor
    return []

# Função principal
def main(tamanho, caso):
    locale.setlocale(locale.LC_ALL, "Portuguese")

    metricas = Metricas()
    arr = carregarVetor(tamanho, caso)
    
    inicio = time.time()
    quickSort(arr,0,tamanho-1, metricas)
    fim = time.time()
    metricas.tempoExecucao = fim - inicio

    # Retornar métricas como dicionário
    return {
        "Comparacoes": metricas.comparacoes,
        "Trocas": metricas.trocas,
        "Tempo de execucao": metricas.tempoExecucao,
        "Memoria usada": metricas.memoriaUsada
    }

if __name__ == "__main__":
    main()

File: Interpretadas/Javascript/quick_sort.js
const fs = require('fs');
const { performance } = require('perf_hooks');

class Metricas {
    constructor() {
        this.comparacoes = 0;
        this.trocas = 0;
        this.tempoExecucao = 0.0; // Em segundos
        this.memoriaUsada = 0; // Em bytes
    }
}

function swap(arr, i, j) {
    [arr[i], arr[j]] = [arr[j], arr[i]];
}

// Função para escolher a mediana de três como pivot
function medianaDeTres(arr, low, high) {
    const mid = Math.floor((low + high) / 2);
    if (arr[low] > arr[mid]) swap(arr, low, mid);
    if (arr[low] > arr[high]) swap(arr, low, high);
    if (arr[mid] > arr[high]) swap(arr, mid, high);
    swap(arr, mid, high); // Coloca o pivot no final
    return arr[high];
}

function partition(arr, low, high, metrics) {
    const pivot = medianaDeTres(arr, low, high);
    metrics.comparacoes += 3;
    let i = (low - 1);
    metrics.memoriaUsada += (pivot+i)*4;
    for (let j = low; j < high; j++) {
        metrics.comparacoes++;
        if (arr[j] < pivot) {
            i++;
            swap(arr, i, j);
            metrics.trocas++;
        }
    }
    swap(arr, i + 1, high);
    metrics.trocas++;
    return (i + 1);
}

function quickSort(arr, low, high, metrics) {
    if (low < high) {
        const pi = partition(arr, low, high, metrics);

        quickSort(arr, low, pi - 1, metrics);
        quickSort(arr, pi + 1, high, metrics);
    }
}

function carregarVetor(tamanho, caso, caminhoVetor) {
    const data = fs.readFileSync(caminhoVetor, 'utf8');
    const lines = data.split('\n');
    const key = `${tamanho} ${caso}:`;
    for (let i = 0; i < lines.length; i++) {
        if (lines[i].trim() === key) {
            return lines[i + 1].trim().split(' ').map(Number);
        }
    }
    return [];
}

function main(tamanho, caso, caminhoVetor) {
    const metricas = new Metricas();
    const arr = carregarVetor(tamanho, caso, caminhoVetor);
    
    const inicio = performance.now();
    quickSort(arr, 0, arr.length - 1, metricas);
    const fim = performance.now();
    
    metricas.tempoExecucao = (fim - inicio) / 1000; // Em segundos
    return {
        "Comparacoes": metricas.comparacoes,
        "Trocas": metricas.trocas,
        "Tempo de execucao": metricas.tempoExecucao,
        "Memoria usada": metricas.memoriaUsada
    };
}

const args = process.argv.slice(2);
const tamanho = parseInt(args[0]);
const caso = args[1];
const caminhoVetor = args[2];

const resultado = main(tamanho, caso, caminhoVetor);
console.log(resultado);